### ISSUE TITLE
FineUploader 5.5.1 multipart version 4 signature bug with unicode filenames

### ISSUE THREAD

**Issue by iliaznk** [FineUploader/fine-uploader#1553]
## Type of issue
Bug

## Uploader type
S3

### Fine Uploader version
5.5.1

### Browsers where the bug is reproducible
Chrome 49.0.2623.110 (64-bit)

### Operating systems where the bug is reproducible
OSX 10.11.4

### Exact steps required to reproduce the issue
Just try to upload a file larger than 5MB with unicode characters in its filename.

### Detailed explanation of the problem
You'll get `"Problem initiating upload request."` and `"The request signature we calculated does not match the signature you provided. Check your key and signing method."`

The problem is that if my filename is, for example, `абвг.jpg` then the 2nd line of the canonical request is encoded as `/%u0430%u0431%u0432%u0433.jpg`, but should be `/%D0%B0%D0%B1%D0%B2%D0%B3.jpg` in which case multipart is initiated properly and the file is uploaded without any errors.

**Comment by iliaznk** [comment]
I modified the last line of `v4.getCanonicalUri` method as `return ("/" + path)` and it seems to be working fine.

**Comment by rnicholus** [comment]
Everything seems to be working fine for me with a 30MB file with that exact name. Most likely there is either an issue server-side on your end, or perhaps your page is missing a `<meta charset="utf-8">` tag. I was using the 1.1.0 version of the [Fine Uploader S3 PHP server code](https://github.com/FineUploader/php-s3-server) on my end (which is currently the latest).

**Comment by iliaznk** [comment]
My page has a `<meta charset="utf-8">` tag and I don't think it has something to do with the server side since this is the data generated by Fine Uploader and sent to the server for signature.

Well, anyway, I got it working for me.

**Comment by rnicholus** [comment]
Since you have a meta tag, then the issue does seem to be server-side. I was able to upload a file using the exact name in your issue report without issue (using v4 signature support). Note that I am using Fine Uploader 5.6.0, so perhaps there is an unreported bug in the version you are using? Upgrading may shed some more light on this.

**Comment by iliaznk** [comment]
But it is data coming from Fine Uploader, how can it be server side? As I mentioned, I changed the return value of `v4.getCanonicalUri` from `return escape("/" + decodeURIComponent(path));` to `return ("/" + path);` to get it working. I also will try to upgrade to the latest version and see if it works there without any modifications.

Thanks for you help!

**Comment by rnicholus** [comment]
I guess I'm not certain if the issue is the data sent by Fine Uploader, or the signature generated by the server. If you are still unable to upload the file after upgrading, then there is most definitely something specific to your environment that is resulting in the error. The tough part may be uncovering that.

**Comment by iliaznk** [comment]
So, I've upgraded and everything is the same, and why wouldn't it when the `v4.getCanonicalUri` method is the same. Just to make it clear, for example, if I have a file named `ф.jpg`, Fine Uploader encodes it in the canonical request as `/%u0444.jpg`, does it look like a normal encoding to you? Those percent signs together with `u`? But everything works fine if I change encoding to `%D1%84.jpg`.

On my server side I just use standard Python 2.7 library to generate the signature.

By the way, in the canonical request I receive from Fine Uploader there's another line with an encoded filename `x-amz-meta-qqfilename:%D1%84.jpg` and as you see here the filename is encoded differently and correctly.

I guess, there's only one way we can check what the problem is: if you could check the canonical request you get from Fine Uploader in your environment and if a unicode filename is also different in those two line of the canonical request but it's working fine, then.... what does it mean then? :)

**Comment by rnicholus** [comment]
The escape sequence generated by Fien Uploader that you have posted for that filename is correct. The canonical URI is not URL encoded, it is escaped. The process required to properly handle v4 signatures is described more in #1336, but this logic appears to be required based on discussions at the end of that issue, and is handled correctly by the example PHP, Java, and Node.JS signature servers maintained in this organization. So it seems the issue is indeed with your server code. 

Properly encoding requests to satisfy AWS is a continual source of headaches, as you can see.

**Comment by iliaznk** [comment]
I see. Well, at least I made it work and AWS is happy with my signature. Thanks again!

**Comment by rnicholus** [comment]
Did this involve changing internal Fine Uploader code? If so, we'll want to find a better solution as this will make upgrading a problem for you.

**Comment by iliaznk** [comment]
Yes, it did. I don't personally see it as a major problem. I'll just modify it again, when I have to. I've made some small modifications to the Fine Uploader code to suite my needs before that anyway.

By the way, this is what [Wikipedia](https://en.wikipedia.org/wiki/Percent-encoding#Non-standard_implementations) says:

> There exists a non-standard encoding for Unicode characters: %uxxxx, where xxxx is a UTF-16 code unit represented as four hexadecimal digits. This behavior is not specified by any RFC and has been rejected by the W3C. The third edition of ECMA-262 still includes an escape function that uses this syntax, along with encodeURI and encodeURIComponent functions, which apply UTF-8 encoding to a string, then percent-escape the resulting bytes.

**Comment by rnicholus** [comment]
Yes, that is my understanding as well. I was unaware of a better way to address the AWS-specific requirements that came out while working on #1336. I believe AWS' JS SDK made use of `escape` as well, but perhaps there is a better solution. I'll keep this case open to remind myself to look into this again.

**Comment by iliaznk** [comment]
Sounds great, thanks!
